Add company referral activity stream endpoint.

elcct opened the pull request:
### Description of change

This adds `/v3/activity-stream/company-referral` endpoint that lists company referrals. This endpoint will be consumed by the activity stream.

Company referral sender is an `actor`, recipient is a `target` and if referral is closed or completed, the relevant person will be appended to `attributedTo` list.

### Checklist

* [x] Has a new newsfragment been created? Check [changelog/README.md](https://github.com/uktrade/data-hub-api/blob/master/changelog/README.md) for instructions
* [ ] Do any added or updated endpoints appear in the API documentation? See [docs/Maintaining the API documentation.md](https://github.com/uktrade/data-hub-api/blob/develop/docs/Maintaining&#32;the&#32;API&#32;documentation.md) for more details
* [ ] Have any relevant search models been updated?
* [ ] Have any relevant fixtures (`fixtures/test_data.yaml`) been updated?
* [x] Have any relevant select-/prefetch-related field lists in the views and search apps been updated?
* [ ] Has the admin site been updated (for new models, fields etc.)?
* [ ] Has the README been updated (if needed)?

reupen commented:
@elcct Some of the tests failed on CircleCI.

reupen commented on datahub/activity_stream/company_referral/serializers.py:
```diff
@@ -0,0 +1,61 @@
+                'id': company_referral_id,
+                'dit:subject': instance.subject,
+                'dit:status': str(instance.status),
+                'dit:notes': instance.notes,
```
Do we need `notes` in here? They might be quite large.

elcct replied:
You're right. I was having doubts about it too. I have removed it now. Thank you!

reupen commented on datahub/activity_stream/company_referral/serializers.py:
```diff
@@ -0,0 +1,61 @@
+            'object': {
+                'id': company_referral_id,
+                'dit:subject': instance.subject,
+                'dit:status': str(instance.status),
```
[We previously mapped interaction kinds](https://github.com/uktrade/data-hub-api/blob/c48ac34f3acf1555fbcac5d652498db5748acf20/datahub/activity_stream/interaction/serializers.py#L8-L11), I believe to be compliant with https://www.w3.org/TR/activitystreams-core/. Do we need to do that here? I think @alixedi or @michalc might be able to confirm if we do or if it actually matters. (I think if we can avoid it, it would probably keep things simpler.)

reupen replied:
Actually, those were in `type`, so it doesn't look applicable here.

reupen commented on datahub/activity_stream/company_referral/views.py:
```diff
@@ -0,0 +1,27 @@
+
+    pagination_class = CompanyReferralCursorPagination
+    serializer_class = CompanyReferralActivitySerializer
+    queryset = CompanyReferral.objects
```
Would be good to use `select_related` to reduce the number of queries.

elcct replied:
Yes! I forgot about this. It is now added. Thank you!

elcct commented:
@reupen yes, I had a typo in one of the apps, but it is now fixed. Thanks!

reupen commented on datahub/activity_stream/company_referral/serializers.py:
```diff
@@ -0,0 +1,48 @@
+            company_referral['object']['dit:closed_on'] = instance.closed_on
+            company_referral['object']['attributedTo'].append(
+                self._get_adviser(instance.closed_by),
+            )
```
Is putting `completed_by` and `closed_by` in `attributedTo` clear enough for the front end?

Also, it seems this will affect how the 'My activity' filter works: https://github.com/uktrade/data-hub-frontend/blob/79a80fd4a5a3737367c74e7778d5456d1a521e96/src/apps/companies/apps/activity-feed/builders.js#L35-L39

Do we know which referrals should show when the 'My activity' filter is used?

elcct replied:
I was thinking about giving it under `dit:completed_by` etc. or adding another object to the collection with e.g.:

```
{
      "type": "Complete",
      "actor": { <adviser data> },
      "object": <company referral ID>
}
```

but I am not sure if that would be too complex.
Having the adviser who either completed or closed in `attributedTo` was fine to FE dev. 
What do you think?

reupen replied:
I think we need to find out the answer to:

> Do we know which referrals should show when the 'My activity' filter is used?

first.

elcct replied:
It looks like we should show all items that the adviser is involved in.

reupen replied:
So, if I'm not mistaken, that currently means all of `created_by`, `recipient`, `completed_by` and `closed_by` should be in `attributedTo`...

elcct replied:
I have updated the format so all advisers are in `attributedTo` and the information about performed action is stored in the `context`.

reupen commented on datahub/activity_stream/company_referral/serializers.py:
```diff
@@ -0,0 +1,48 @@
+            'generator': self._get_generator(),
+            'actor': self._get_adviser(instance.created_by),
+            'target': self._get_adviser(instance.recipient),
+            'object': {
```
I think we need `startTime` in `object` as well as it seems to be what is used for sorting: https://github.com/uktrade/data-hub-frontend/blob/79a80fd4a5a3737367c74e7778d5456d1a521e96/src/apps/companies/apps/activity-feed/builders.js#L28

elcct replied:
Good catch! I have added it now. Thank you!

reupen commented on datahub/activity_stream/company_referral/serializers.py:
```diff
@@ -0,0 +1,59 @@
+                'dit:status': str(instance.status),
+                'attributedTo': [
+                    self._get_company(instance.company),
+                    self._get_contact(instance.contact),
```
`contact` is optional, it missing doesn't seem to be handled at the moment.

elcct replied:
That's true. It is now checked I also added a test. Thank you!

reupen commented on datahub/activity_stream/company_referral/views.py:
```diff
@@ -0,0 +1,32 @@
+        'contact',
+        'created_by',
+        'recipient',
+    )
```
Would be good to add `completed_by` and `closed_by` too.

elcct replied:
I added it now and also for `dit_team`. Thank you!

reupen commented:
@elcct Could you explain why we decided to make `context` an object with a timestamp instead of something simpler?

Also, don't we need team info with each the adviser as well?

elcct commented:
@reupen I thought that this information gives a context into why particular adviser is included in the list and I thought this will be easier to process when such object attached to adviser object. I initially wanted to include that information at the referral object level, but I thought this will complicate parsing. If you have better way in mind, I am happy to change it.

Regarding teams, I was too focused on general shape of data I missed this detail. We indeed need this information, so I have added it now. Thank you!

reupen commented on datahub/activity_stream/company_referral/views.py:
```diff
@@ -0,0 +1,38 @@
+        'company',
+        'contact',
+        'created_by',
+        'created_by__dit_team',
```
`created_by__dit_team` implies `created_by`, so we could get rid of `created_by` etc.

elcct replied:
That is true! Updated. Thank you.

reupen commented on datahub/activity_stream/company_referral/serializers.py:
```diff
@@ -0,0 +1,74 @@
+
+        company_referral_id = f'dit:DataHubCompanyReferral:{instance.pk}'
+        company_referral = {
+            'id': f'{company_referral_id}:{activity_type}',
```
I'm not sure about varying the ID like this depending on status. Do we have a risk of duplicates until the activity stream service completes a full re-index?

elcct replied:
As per our conversation I understand this will cause unintended behaviour, so I removed this change. Thank you!

elcct commented:
Your proposed way will make it even simpler to parse, so I replaced the context with a role and moved the date of either completion or closure to the referral object. I am not sure whether the role would be generic enough, so I namespaced it. Thank you!

currycoder commented on datahub/activity_stream/company_referral/views.py:
```diff
@@ -0,0 +1,34 @@
+    """
+    Cursor pagination for Company Referral.
+
+    `modified_on` is no unchanging but we have decided to use it because the benefits of being
```
Could this read better?

"`modified_on` is mutable"

currycoder replied:
What would the alternative have been which would have meant 10s queries?  Might be good to fill in the blanks for readers of this comment who don't have all the context.

elcct replied:
This is actually a copy & paste comment from `InteractionCursorPagination` and I missed that. I have changed it so it describes why `modified_on` is being used, I hope it makes more sense now.
